### PR TITLE
add a comparison of classic and modern polymorphism

### PR DESCRIPTION
**Pull request by daixtrose**
## Summary by CodeRabbit

- **New Features**
	- Introduced a comprehensive example demonstrating C++ polymorphism with both classic and modern approaches.
	- Added a detailed README file outlining usage, implementation methods, and build instructions.
	- Included a new interface for polymorphism and its implementation.
	- Enhanced build workflows to support parallel building and testing configurations.

- **Bug Fixes**
	- Improved reliability and output clarity in build and test workflows.

- **Documentation**
	- Updated README to include detailed explanations of polymorphism methodologies and testing instructions.

- **Tests**
	- Added unit tests to ensure the correctness of both classic and modern interface implementations.
	- Introduced a new CMake configuration for testing with the Boost Extended Unit Testing library.
	- Enabled testing options in the project configuration.
	- Added a new CMake configuration for the test directory.

### CODE REVIEW

File: polymorphism/test/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+cmake_minimum_required(VERSION 3.29)
+
+include(FetchContent)
+
+FetchContent_Declare(ut
+    URL https://raw.githubusercontent.com/boost-ext/ut/refs/heads/master/include/boost/ut.hpp
+    DOWNLOAD_NO_EXTRACT TRUE
+    SOURCE_DIR ut
+)
+FetchContent_MakeAvailable(ut)
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/test/CMakeLists.txt:10]: I want to use the latest version here, so forget about this issue.

File: polymorphism/test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+add_executable(test_consume
+    test_consume.cpp
+    ../src/consume_class_that_adheres_to_concept.cpp
+    ../src/consume_class_with_interface.cpp
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/test/CMakeLists.txt]: Fixed in 67cab71be24962b038954b69f1483ee827b8a0db

File: .github/workflows/build_examples.yml [hunk 3]
```diff
@@ -40,6 +40,12 @@ jobs:
             build_example_3/${{ steps.check-packages.outputs.DEBIAN_PACKAGE_FILENAME }}
           token: ${{ secrets.GITHUB_TOKEN }}
           tag_name: latest
+      - name: Build and test polymorphism
+        run: |
+          cmake -D CMAKE_BUILD_TYPE=Debug -B build_polymorphism -S polymorphism
+          VERBOSE=1 cmake --build build_polymorphism --clean-first --config Debug --target all --parallel
+          ctest --test-dir build_polymorphism  --output-on-failure --build-config Debug
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply .github/workflows/build_examples.yml]: fixed in 5ecc7d1507ea17905188ce9a2f40bb10ba47b85b

File: polymorphism/include/polymorphism/impl_with_interface.hpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+public:
+    std::string coolFeature() const override { return s_; }
+    void set(std::string s) override
+    {
+        s_ = std::move(s);
+    }
+};
+
+} // namespace modern
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/include/polymorphism/impl_with_interface.hpp]: fixed

File: polymorphism/src/consume_class_that_adheres_to_concept.cpp [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#include <consume_class_that_adheres_to_concept.ipp>
+#include <polymorphism/impl_without_interface.hpp>
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/src/consume_class_that_adheres_to_concept.cpp:2]: fixed in 850b2eba6b8854c811791c34b873a73f05501d87

File: polymorphism/src/consume_class_that_adheres_to_concept.cpp [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+
+namespace modern {
+
+// Explicit template instatiation
+template std::string consume<Impl>(Impl&);
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/src/consume_class_that_adheres_to_concept.cpp]: fixed in 533f18f68bb0bedfc06d9496db4fb9792c541744

File: polymorphism/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+project(polymorphism VERSION 1.0.0 LANGUAGES CXX)
+
+option(POLYMORPHISM_ENABLE_TESTING "enable testing for polymorphism" ${PROJECT_IS_TOP_LEVEL})
+
+add_executable(${PROJECT_NAME}
+    src/consume_class_with_interface.cpp
+    src/consume_class_that_adheres_to_concept.cpp
+    src/main.cpp
+)
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/CMakeLists.txt:12]: For this demo project with only 3 files to compile I am fine with my approach. Do not complain any longer.

File: polymorphism/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,20 @@
+cmake_minimum_required(VERSION 3.29)
+
+project(polymorphism VERSION 1.0.0 LANGUAGES CXX)
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/CMakeLists.txt:3]: Please remember that using the latest and greatest C++ and CMake features is **on purpose** for this project.

File: polymorphism/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+add_executable(${PROJECT_NAME}
+    src/consume_class_with_interface.cpp
+    src/consume_class_that_adheres_to_concept.cpp
+    src/main.cpp
+)
+
+target_compile_features(${PROJECT_NAME} PUBLIC cxx_std_23)
+target_include_directories(${PROJECT_NAME} PUBLIC include)
+target_include_directories(${PROJECT_NAME} PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/src)
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/CMakeLists.txt]: fixed

File: polymorphism/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+if(POLYMORPHISM_ENABLE_TESTING)
+    enable_testing()
+    add_subdirectory(test)
+endif()
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/CMakeLists.txt]: proposal applied

File: polymorphism/test/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+target_include_directories(test_consume PUBLIC ../include)
+target_include_directories(test_consume PUBLIC "${ut_SOURCE_DIR}")
+
+target_link_libraries(test_consume PRIVATE polymorphism_lib)
+
+add_test(
+    NAME test_consume
+    COMMAND test_consume
+)
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/test/CMakeLists.txt]: A WORKING_DIRECTORY has been added in ff39cadf25a78dbdf53062743328f9827e2cfa0a, but the TIMEOUT option is unavailable. The reference documentation at https://cmake.org/cmake/help/latest/command/add_test.html does not mention it. Citing from the docs:

```cmake
add_test(NAME <name> COMMAND <command> [<arg>...]
         [CONFIGURATIONS <config>...]
         [WORKING_DIRECTORY <dir>]
         [COMMAND_EXPAND_LISTS])
```

So, I'm sorry, you are in error here.

File: polymorphism/README.md [hunk 12]
```diff
@@ -0,0 +1,192 @@
+## Summary
+
+This example shows two different ways to implement [polymorphism](https://en.wikipedia.org/wiki/Polymorphism_(computer_science)) in C++:
+
+1. The classic approach using an [abstract class](https://en.cppreference.com/w/cpp/language/abstract_class) which is used to define an interface and from which the implementation inherits.
+2. The modern approach using [concepts](https://en.cppreference.com/w/cpp/language/constraints) to declare features a type must have. 
+
+In addition, this directory contains an example about how to perform unit tests using the [ut testing framework](https://github.com/boost-ext/ut).
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/README.md]: fixed.

File: polymorphism/README.md [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+```
+
+
+### Building the `polymorphism` project
+
+Call `cmake` from the top level like this:
+
+```bash
+cmake --build build_polymorphism --config Release --target polymorphism --parallel
+```
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/README.md]: fixed

File: polymorphism/README.md [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+
+You can find the definition of `consume` in `src/consume_class_that_adheres_to_concept.ipp` which is then included into `src/consume_class_that_adheres_to_concept.cpp`. In the latter file we also place all explicit instantiations of `consume`:
+
+```c++
+// explicit template instantiation
+template std::string consume<Impl>(Impl&);
+```
+
+Als an alternative, the *definition* of `consume` could be in the header `include/polymorphism/consume_class_that_adheres_to_concept.hpp`. Then the explicit template instantiation is no longer required. 
+
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/README.md]: fixed in 38aa34efc9b1331c646c2528b1f5bf58634b0390

File: polymorphism/README.md [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+FetchContent_MakeAvailable(ut)
+```
+
+and the directory to which the header is downloaded is added to the search path set for the build of the test executables: 
+
+```cmake
+target_include_directories(test_consume PUBLIC "${ut_SOURCE_DIR}")
+```
+
+Please refer to [this slidedeck](https://boost-ext.github.io/ut/denver-cpp-2019/#/) for further documentation of this test library.
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/README.md]: fixed in e0d3d98fbd04ef8cc01ba722918c622ae73b9771

[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/README.md]: I added a brief explanation of why µt was chosen in 8cad8a1e50eeee5adfd5f0b43f7559a5df8af9c0

File: polymorphism/test/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+FetchContent_MakeAvailable(ut)
+
+add_library(polymorphism_lib
+    ../src/consume_class_that_adheres_to_concept.cpp
+    ../src/consume_class_with_interface.cpp
+)
+
+target_compile_features(polymorphism_lib PUBLIC cxx_std_23)
+target_include_directories(polymorphism_lib PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/../include)
+target_include_directories(polymorphism_lib PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/../src)
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/test/CMakeLists.txt]: Your proposal will not work. We need the `*.ipp` file to be found.

**Comment by daixtrose** [comment]
@coderabbitai, please recheck the pull request and summarize your findings.

File: polymorphism/test/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+
+add_executable(test_consume
+    test_consume.cpp
+    ../src/consume_class_that_adheres_to_concept.cpp
+    ../src/consume_class_with_interface.cpp
+)
+
+target_compile_features(test_consume PUBLIC cxx_std_23)
+target_include_directories(test_consume PUBLIC ../include)
+target_include_directories(test_consume PUBLIC "${ut_SOURCE_DIR}")
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/test/CMakeLists.txt]: I want to use the greatest and latest compiler and CMake for this project.

**Comment by daixtrose** [comment]
@coderabbitai, please recheck the pull request and summarize your findings. Please respect in your analysis that I want to use the latest versions of all build and configuration tools, especially for compilers, C++ versions, and CMake.

File: polymorphism/include/polymorphism/consume_class_that_adheres_to_concept.hpp [hunk 18]
```diff
@@ -0,0 +1,13 @@
+#ifndef POYMORPHISM_CONSUME_CLASS_THAT_ADHERES_TO_CONCEPT_HPP
+#define POYMORPHISM_CONSUME_CLASS_THAT_ADHERES_TO_CONCEPT_HPP
+
+#include <polymorphism/has_super_cool_features.hpp>
+
+namespace modern {
+
+// declaration with concept constraint
+std::string consume(has_super_cool_features auto& s);
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/include/polymorphism/consume_class_that_adheres_to_concept.hpp:11]: @coderabbitai please provide a committable suggestion

**Comment by daixtrose** [comment]
@coderabbitai, please create code change suggestions that add `CPLUSPLUS_PRIMER_` to all include guards in files below the directory `polymorphism`.

File: polymorphism/include/polymorphism/i_super_cool_features.hpp [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace classic {
+
+/// @brief A typical interface definition via a class with pure virtual functions/methods.
+class ISuperCoolFeatures {
+public:
+    [[nodiscard]] virtual std::string coolFeature() const = 0; // Pure virtual function
+    virtual void set(std::string s) = 0; // Pure virtual function
+    virtual ~ISuperCoolFeatures() = default; // Virtual destructor
+};
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/include/polymorphism/i_super_cool_features.hpp]: fixed in d9d524a674255f555cf194eebf8793bad3c7fdf9

File: polymorphism/include/polymorphism/impl_without_interface.hpp [hunk 20]
```diff
@@ -0,0 +1,27 @@
+#ifndef POYMORPHISM_IMPL_WITHOUT_INTERFACE_HPP
+#define POYMORPHISM_IMPL_WITHOUT_INTERFACE_HPP
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/include/polymorphism/impl_without_interface.hpp]: fixed

File: polymorphism/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,23 @@
+cmake_minimum_required(VERSION 3.29)
+
+project(polymorphism VERSION 1.0.0 LANGUAGES CXX)
+
+option(POLYMORPHISM_ENABLE_TESTING "enable testing for polymorphism" ${PROJECT_IS_TOP_LEVEL})
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/CMakeLists.txt:6]: done

File: polymorphism/include/polymorphism/impl_without_interface.hpp [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+public:
+    std::string coolFeature() const { return s_; }
+    void set(std::string s)
+    {
+        s_ = std::move(s);
+    }
+};
+
+// Check if the class adheres to the concept (i.e. has the interface we want it to have).
+static_assert(has_super_cool_features<Impl>);
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/include/polymorphism/impl_without_interface.hpp]: done

File: polymorphism/test/CMakeLists.txt [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+target_include_directories(polymorphism_lib PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/../src)
+
+add_executable(test_consume
+    test_consume.cpp
+)
+
+target_compile_features(test_consume PUBLIC cxx_std_23)
+
+target_include_directories(test_consume PUBLIC ../include)
+target_include_directories(test_consume PUBLIC "${ut_SOURCE_DIR}")
```
[the remark this answers is not in the document]
Comment (daixtrose) [reply polymorphism/test/CMakeLists.txt:28]: no

**Comment by daixtrose** [comment]
@coderabbitai, thanks for your code change suggestions. Can you please provide them so that I can click on them and they will be immediately applied?

File: polymorphism/CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (TorstenRobitzki) [reply polymorphism/CMakeLists.txt:12]: What would be the point to use a variable for a list of files, that is going to be used exactly once?

**Pull request merged**